[Liquid Glass] theguardian.com: scroll edge effect makes crossword puzzle hints unreadable on some iPhone models

whsieh opened the pull request:
#### 5f9de3d7d20770fedf23d17e2471795515914eb9
<pre>
[Liquid Glass] theguardian.com: scroll edge effect makes crossword puzzle hints unreadable on some iPhone models
<a href="https://bugs.webkit.org/show_bug.cgi?id=295613">https://bugs.webkit.org/show_bug.cgi?id=295613</a>
<a href="https://rdar.apple.com/153818837">rdar://153818837</a>

Reviewed by Aditya Keerthi.

Add a site-specific quirk on `theguardian.*` to hide the top scroll edge effect when focusing on
input fields in a crossword puzzle. On smaller phones, this prevents the top scroll edge effect&apos;s
variable blur from extending down and overlapping the top hint text.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldHideSoftTopScrollEdgeEffectDuringFocus const):
(WebCore::handleGuardianQuirks):

Add support for the new quirk, based on the ID of the focused element.

(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebKit/Shared/FocusedElementInformation.h:
* Source/WebKit/Shared/FocusedElementInformation.serialization.in:

Add a flag that stores whether or not the top scroll pocket should be hidden when focusing this
element.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Add a new reason to hide the top scroll pocket — `SiteSpecificQuirk`. One flag should suffice for
now, assuming we don&apos;t end up with multiple site-specific quirks that need to be active at the same
time, on the same site.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateNeedsTopScrollPocketDueToVisibleContentInset]):
* Source/WebKit/UIProcess/ios/UIKitUtilities.h:
* Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
(-[UIScrollView _wk_usesHardTopScrollEdgeEffect]):

Add a helper method to return whether or not the top scroll edge effect uses the hard scroll edge
effect style.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _continueElementDidFocus:requiresStrongPasswordAssistance:focusedElementInfo:activityStateChanges:restoreValues:]):
(-[WKContentView _elementDidBlur]):

Implement the main fix here, by adding or removing the new `SiteSpecificQuirk` reason based on the
value of `shouldHideSoftTopScrollEdgeEffect` from the incoming focused element info. We only remove
the hiding reason here if `!_isChangingFocus`, to prevent the reason flag from being thrashed
(removed and immediately re-introduced) in the case where we&apos;re moving from one quirked text input
to another.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::focusedElementInformation):

Canonical link: <a href="https://commits.webkit.org/297153@main">https://commits.webkit.org/297153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7af6148f547840334b1eb7d221a0aa8abc8266e3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110700 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30359 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20794 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116727 "Built successfully") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60967 "Failed to checkout and rebase branch from PR 47749") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112663 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31038 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38948 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84176 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/60967 "Failed to checkout and rebase branch from PR 47749") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113648 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24795 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99685 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64617 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24160 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17823 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60521 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94185 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17882 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119517 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37740 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28043 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93138 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38114 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95953 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92962 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23690 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37994 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15740 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33718 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37636 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43108 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37298 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40637 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39006 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -8542,6 +8549,11 @@ - (void)_elementDidBlur
     [self _endEditing];
 
+#if ENABLE(CONTENT_INSET_BACKGROUND_FILL)
+    if (!_isChangingFocus)
```
It's not immediately clear why `!_isChangingFocus` needs to be checked, can you explain?

whsieh replied:
Ah, the `!_isChangingFocus` check is here to prevent `HideScrollPocketReason::SiteSpecificQuirk` from being removed and added again in the case where you're moving from one focused input to another (both of which would trigger this quirk).

I can clarify this in the commit message.

whsieh committed:
[Liquid Glass] theguardian.com: scroll edge effect makes crossword puzzle hints unreadable on some iPhone models
https://bugs.webkit.org/show_bug.cgi?id=295613
rdar://153818837

Reviewed by Aditya Keerthi.

Add a site-specific quirk on `theguardian.*` to hide the top scroll edge effect when focusing on
input fields in a crossword puzzle. On smaller phones, this prevents the top scroll edge effect's
variable blur from extending down and overlapping the top hint text.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldHideSoftTopScrollEdgeEffectDuringFocus const):
(WebCore::handleGuardianQuirks):

Add support for the new quirk, based on the ID of the focused element.

(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebKit/Shared/FocusedElementInformation.h:
* Source/WebKit/Shared/FocusedElementInformation.serialization.in:

Add a flag that stores whether or not the top scroll pocket should be hidden when focusing this
element.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Add a new reason to hide the top scroll pocket — `SiteSpecificQuirk`. One flag should suffice for
now, assuming we don't end up with multiple site-specific quirks that need to be active at the same
time, on the same site.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateNeedsTopScrollPocketDueToVisibleContentInset]):
* Source/WebKit/UIProcess/ios/UIKitUtilities.h:
* Source/WebKit/UIProcess/ios/UIKitUtilities.mm:
(-[UIScrollView _wk_usesHardTopScrollEdgeEffect]):

Add a helper method to return whether or not the top scroll edge effect uses the hard scroll edge
effect style.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _continueElementDidFocus:requiresStrongPasswordAssistance:focusedElementInfo:activityStateChanges:restoreValues:]):
(-[WKContentView _elementDidBlur]):

Implement the main fix here, by adding or removing the new `SiteSpecificQuirk` reason based on the
value of `shouldHideSoftTopScrollEdgeEffect` from the incoming focused element info. We only remove
the hiding reason here if `!_isChangingFocus`, to prevent the reason flag from being thrashed
(removed and immediately re-introduced) in the case where we're moving from one quirked text input
to another.

* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::focusedElementInformation):

Canonical link: https://commits.webkit.org/297153@main
